File: src/widgets/git_status.zig
const std = @import("std");
const tmux_renderer = @import("../tmux_renderer.zig");
const util = @import("../core/util.zig");
const WidgetContext = @import("../core/widget.zig").WidgetContext;
const runGitCommand = util.runGitCommand;
const trimBranchName = util.trimBranchName;

const SyncMode = enum {
    clean,
    dirty,
    ahead,
    behind,
};

fn getPorcelainV2(allocator: std.mem.Allocator, io: std.Io, repo_path: []const u8) !util.ParsedStatusV2 {
    const stdout = runGitCommand(allocator, io, &.{ "git", "status", "--porcelain=v2", "--branch" }, repo_path) catch {
        return util.ParsedStatusV2{
            .branch = null,
            .ahead = 0,
            .behind = 0,
            .changed = 0,
            .untracked = 0,
            .conflicts = 0,
        };
    };
    defer allocator.free(stdout);

    var status = util.parsePorcelainV2(stdout);

    // Dupe the branch name — parsePorcelainV2 returns a slice into stdout
    // which will be freed by the defer above. All other fields are integers.
    if (status.branch) |branch| {
        status.branch = try allocator.dupe(u8, branch);
    }

    return status;
}

fn getDiffStats(allocator: std.mem.Allocator, io: std.Io, repo_path: []const u8) !struct { changed: usize, insertions: usize, deletions: usize } {
    const stdout = runGitCommand(allocator, io, &.{ "git", "diff", "--numstat", "HEAD" }, repo_path) catch return .{ .changed = 0, .insertions = 0, .deletions = 0 };
    defer allocator.free(stdout);

    var changed: usize = 0;
    var insertions: usize = 0;
    var deletions: usize = 0;

    var lines = std.mem.splitScalar(u8, stdout, '\n');
    while (lines.next()) |line| {
        var it = std.mem.splitScalar(u8, line, '\t');
        const ins_str = it.next() orelse continue;
        const del_str = it.next() orelse continue;
        if (ins_str.len == 0 or del_str.len == 0) continue;

        const ins = std.fmt.parseInt(usize, ins_str, 10) catch 0;
        const del = std.fmt.parseInt(usize, del_str, 10) catch 0;
        changed += 1;
        insertions += ins;
        deletions += del;
    }

    return .{ .changed = changed, .insertions = insertions, .deletions = deletions };
}

fn countStashes(allocator: std.mem.Allocator, io: std.Io, repo_path: []const u8) !usize {
    const stdout = runGitCommand(allocator, io, &.{ "git", "rev-list", "--count", "refs/stash" }, repo_path) catch return 0;
    defer allocator.free(stdout);

    const trimmed = std.mem.trim(u8, stdout, " \n\r");
    return std.fmt.parseInt(usize, trimmed, 10) catch 0;
}

pub fn run(
    allocator: std.mem.Allocator,
    io: std.Io,
    environ_map: *std.process.Environ.Map,
    theme_name: []const u8,
    transparent: bool,
    repo_path: []const u8,
    writer: *std.Io.Writer,
) !void {
    var ctx = try WidgetContext.init(allocator, io, environ_map, theme_name, transparent);
    defer ctx.deinit();
    const theme = ctx.theme;
    const reset = ctx.reset;

    // Single call replaces: rev-parse, status --porcelain, rev-list --count, diff --name-only --diff-filter=U, stash list
    const status = try getPorcelainV2(allocator, io, repo_path);

    const branch_raw = status.branch orelse return;
    defer allocator.free(branch_raw);
    if (branch_raw.len == 0 or std.mem.eql(u8, branch_raw, "HEAD")) return;

    const display_branch = try trimBranchName(allocator, branch_raw);
    defer allocator.free(display_branch);

    const untracked = status.untracked;
    const conflict_count = status.conflicts;
    const ahead = status.ahead;
    const behind = status.behind;
    const changed = status.changed;

    var sync_mode: SyncMode = .clean;

    if (changed > 0) {
        sync_mode = .dirty;
    }

    if (ahead > 0 and sync_mode == .clean) {
        sync_mode = .ahead;
    } else if (behind > 0 and sync_mode == .clean) {
        sync_mode = .behind;
    }

    // Get diff stats and stash count using optimized git commands
    const diff_stats = try getDiffStats(allocator, io, repo_path);
    const stash_count = try countStashes(allocator, io, repo_path);

    const insertions = diff_stats.insertions;
    const deletions = diff_stats.deletions;

    // Pre-allocate color hex strings for common colors to reduce allocations
    // Use stack-allocated buffers for better performance
    var bg_hex: [32]u8 = undefined;
    var warning_hex: [32]u8 = undefined;
    var success_hex: [32]u8 = undefined;
    var danger_hex: [32]u8 = undefined;
    var danger_bright_hex: [32]u8 = undefined;
    var info_bright_hex: [32]u8 = undefined;
    var muted_hex: [32]u8 = undefined;
    const bg_str = tmux_renderer.colorHexString(theme.background, &bg_hex);
    const warning_str = tmux_renderer.colorHexString(theme.warning, &warning_hex);
    const success_str = tmux_renderer.colorHexString(theme.success, &success_hex);
    const danger_str = tmux_renderer.colorHexString(theme.danger, &danger_hex);
    const danger_bright_str = tmux_renderer.colorHexString(theme.danger_bright, &danger_bright_hex);
    const info_bright_str = tmux_renderer.colorHexString(theme.info_bright, &info_bright_hex);
    const muted_str = tmux_renderer.colorHexString(theme.muted, &muted_hex);

    // Write status segments directly to the output writer, avoiding intermediate allocations
    switch (sync_mode) {
        .dirty => try writer.print("{s}#[bg={s},fg={s},bold]▒ 󱓎", .{ reset, bg_str, danger_bright_str }),
        .ahead => try writer.print("{s}#[bg={s},fg={s},bold]▒ 󰛃", .{ reset, bg_str, danger_str }),
        .behind => try writer.print("{s}#[bg={s},fg={s},bold]▒ 󰛀", .{ reset, bg_str, info_bright_str }),
        .clean => try writer.print("{s}#[bg={s},fg={s},bold]▒ ", .{ reset, bg_str, success_str }),
    }

    try writer.print(" {s}{s}", .{ reset, display_branch });

    if (changed > 0) {
        try writer.print(" {s}#[fg={s},bg={s},bold] {d}", .{ reset, warning_str, bg_str, changed });
    }

    if (insertions > 0) {
        try writer.print(" {s}#[fg={s},bg={s},bold] {d}", .{ reset, success_str, bg_str, insertions });
    }

    if (deletions > 0) {
        try writer.print(" {s}#[fg={s},bg={s},bold] {d}", .{ reset, danger_str, bg_str, deletions });
    }

    if (untracked > 0) {
        try writer.print(" {s}#[fg={s},bg={s},bold] {d}", .{ reset, muted_str, bg_str, untracked });
    }

    if (stash_count > 0) {
        try writer.print(" {s}#[fg={s},bg={s},bold] {d}", .{ reset, info_bright_str, bg_str, stash_count });
    }

    if (conflict_count > 0) {
        try writer.print(" {s}#[fg={s},bg={s},bold]󰅘 {d}", .{ reset, danger_bright_str, bg_str, conflict_count });
    }

    if (ahead > 0) {
        try writer.print(" {s}#[fg={s},bg={s},bold]↑{d}", .{ reset, info_bright_str, bg_str, ahead });
    }

    if (behind > 0) {
        try writer.print(" {s}#[fg={s},bg={s},bold]↓{d}", .{ reset, danger_str, bg_str, behind });
    }

    try writer.print(" ", .{});
}

test "git_status renders sync status icons correctly" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var ctx = try WidgetContext.init(gpa, io, &env_map, "hard", false);
    defer ctx.deinit();

    // Verify theme colors are non-empty (basic theme sanity)
    try std.testing.expect(!ctx.theme.success.isDefault());
    try std.testing.expect(!ctx.theme.danger.isDefault());
    try std.testing.expect(!ctx.theme.warning.isDefault());
    try std.testing.expect(!ctx.theme.info_bright.isDefault());
    try std.testing.expect(!ctx.theme.danger_bright.isDefault());
    try std.testing.expect(!ctx.theme.muted.isDefault());
    try std.testing.expect(!ctx.theme.background.isDefault());

    // Verify reset string is properly formatted
    try std.testing.expect(std.mem.startsWith(u8, ctx.reset, "#[fg="));
    try std.testing.expect(std.mem.containsAtLeast(u8, ctx.reset, 1, "nobold"));
}

test "git_status.run produces valid tmux output in project repo" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    // Navigate from test's build-cache directory to the project root.
    // @src().file is relative to build root; go up from src/widgets/ to project root.
    const project_root = comptime std.fs.path.dirname(std.fs.path.dirname(std.fs.path.dirname(@src().file))).?;

    var buf: [2048]u8 = undefined;
    var writer: std.Io.Writer = .fixed(&buf);

    // Run in the project's own git repo (guaranteed to exist during tests)
    run(gpa, io, &env_map, "hard", false, project_root, &writer) catch |err| {
        // Skip if git isn't available or repo state is unexpected
        if (err == error.GitError) return;
        return err;
    };

    const output = std.Io.Writer.buffered(&writer);

    // The output should contain these structural elements for any non-empty repo:
    try std.testing.expect(std.mem.containsAtLeast(u8, output, 1, "#[fg="));
    try std.testing.expect(std.mem.containsAtLeast(u8, output, 1, "#[bg="));
    try std.testing.expect(std.mem.containsAtLeast(u8, output, 1, "▒"));

    // Output ends with a trailing space (tmux status convention)
    try std.testing.expect(output.len > 0);
    try std.testing.expect(output[output.len - 1] == ' ');
}

File: src/widgets/git_status.rs
use std::collections::hash_map::DefaultHasher;
use std::fmt::Write;
use std::fs;
use std::hash::{Hash, Hasher};
use std::path::Path;
use std::sync::OnceLock;

use crate::core::cache::{cache_base, read_fresh_cache, write_cache};
use crate::core::util::{parse_porcelain_v2, run_git_command, trim_branch_name, ParsedStatusV2};
use crate::core::widget::WidgetContext;
use crate::core::Theme;
use crate::tmux_renderer::ThemeHex;

/// Cached HOME so we don't syscall std::env::var("HOME") on every render.
fn home_hash() -> u64 {
    static HOME_HASH: OnceLock<u64> = OnceLock::new();
    *HOME_HASH.get_or_init(|| {
        let home = std::env::var("HOME").unwrap_or_default();
        let mut hasher = DefaultHasher::new();
        home.hash(&mut hasher);
        hasher.finish()
    })
}

enum SyncMode {
    Clean,
    Dirty,
    Ahead,
    Behind,
}

fn get_porcelain_v2_raw(repo_path: &str) -> String {
    let stdout = match run_git_command(
        &[
            "git",
            "--no-optional-locks",
            "status",
            "--porcelain=v2",
            "--branch",
            "--show-stash",
        ],
        Some(Path::new(repo_path)),
    ) {
        Ok(out) => out,
        Err(_) => match run_git_command(
            &[
                "git",
                "--no-optional-locks",
                "status",
                "--porcelain=v2",
                "--branch",
            ],
            Some(Path::new(repo_path)),
        ) {
            Ok(out) => out,
            Err(_) => return String::new(),
        },
    };

    String::from_utf8(stdout).unwrap_or_default()
}

// Same idea as the diff-stat cache below: a cache hit costs zero subprocess
// spawns. Branch name / ahead-behind / stash count can lag up to this TTL
// after a checkout, commit, fetch, or stash push — bounded and consistent
// with the tradeoff already accepted for diff stats.
const STATUS_CACHE_TTL_SECS: u64 = 2;

fn status_cache_path(repo_path: &str) -> String {
    let h = hash_path(repo_path);
    let uh = home_hash();
    format!("{}/flavors-tmux-porcelain-{uh:x}-{h:x}", cache_base())
}

/// Cache the parsed porcelain for a repo for the lifetime of one `status`
/// process. tmux may invoke the same widget twice per refresh (e.g. the
/// same repo path from two panes); without this each call re-spawns git.
/// The TTL check still applies on the disk cache, so staleness is bounded.
use std::cell::RefCell;

thread_local! {
    static PORCELAIN_CACHE: RefCell<Option<(String, ParsedStatusV2)>> = const { RefCell::new(None) };
}

fn get_porcelain_v2(repo_path: &str) -> ParsedStatusV2 {
    let cache_path = status_cache_path(repo_path);

    // In-process hit: same repo, already fetched this process lifetime.
    PORCELAIN_CACHE.with(|c| {
        let mut cache = c.borrow_mut();
        if let Some((path, parsed)) = cache.as_ref() {
            if path == repo_path {
                return parsed.clone();
            }
        }

        let parsed = if let Some(cached) = read_fresh_cache(&cache_path, STATUS_CACHE_TTL_SECS) {
            parse_porcelain_v2(&cached)
        } else {
            let stdout_str = get_porcelain_v2_raw(repo_path);
            write_cache(&cache_path, &stdout_str);
            parse_porcelain_v2(&stdout_str)
        };
        *cache = Some((repo_path.to_owned(), parsed.clone()));
        parsed
    })
}

fn get_diff_stats(repo_path: &str) -> (usize, usize) {
    let stdout = match run_git_command(
        &["git", "--no-optional-locks", "diff", "--numstat", "HEAD"],
        Some(Path::new(repo_path)),
    ) {
        Ok(out) => out,
        Err(_) => return (0, 0),
    };

    let text = String::from_utf8(stdout).unwrap_or_default();
    let mut insertions = 0usize;
    let mut deletions = 0usize;

    for line in text.lines() {
        let mut it = line.split('\t');
        let ins_str = it.next().unwrap_or("");
        let del_str = it.next().unwrap_or("");
        if ins_str.is_empty() || del_str.is_empty() {
            continue;
        }
        insertions += ins_str.parse::<usize>().unwrap_or(0);
        deletions += del_str.parse::<usize>().unwrap_or(0);
    }

    (insertions, deletions)
}

fn hash_path(path: &str) -> u64 {
    let mut hasher = DefaultHasher::new();
    path.hash(&mut hasher);
    hasher.finish()
}

fn diff_cache_path(repo_path: &str) -> String {
    let h = hash_path(repo_path);
    let uh = home_hash();
    format!("{}/flavors-tmux-diff-{uh:x}-{h:x}", cache_base())
}

// Short enough that edits to tracked files (or a new commit) during an
// active session show up within one or two statusline redraws, long enough
// that a cache hit costs zero subprocess spawns instead of one.
//
// The cache key is deliberately just the file's mtime age, not HEAD: reading
// HEAD used to cost its own `git rev-parse` spawn on every single render,
// even on a cache hit, which defeated half the point of caching. The
// tradeoff is that a commit landing inside the TTL window can show
// pre-commit insertion/deletion counts for up to DIFF_CACHE_TTL_SECS —
// bounded and imperceptible at this TTL, same tradeoff already accepted for
// working-tree edits.
const DIFF_CACHE_TTL_SECS: u64 = 3;

fn get_diff_stats_cached(repo_path: &str) -> (usize, usize) {
    let cache_path = diff_cache_path(repo_path);

    let cache_fresh = fs::metadata(&cache_path)
        .ok()
        .and_then(|meta| meta.modified().ok())
        .and_then(|mtime| mtime.elapsed().ok())
        .map(|age| age.as_secs() < DIFF_CACHE_TTL_SECS)
        .unwrap_or(false);

    if cache_fresh {
        if let Ok(content) = fs::read_to_string(&cache_path) {
            let mut parts = content.splitn(2, ' ');
            if let (Some(ins_str), Some(del_str)) = (parts.next(), parts.next()) {
                if let (Ok(ins), Ok(del)) = (ins_str.parse(), del_str.parse()) {
                    return (ins, del);
                }
            }
        }
    }
    let stats = get_diff_stats(repo_path);
    let tmp_path = format!("{cache_path}.tmp");
    let _ = fs::write(&tmp_path, format!("{} {}", stats.0, stats.1));
    let _ = fs::rename(&tmp_path, &cache_path);
    stats
}

/// Fetch + parse the porcelain status for `repo_path`, using the in-process
/// and on-disk caches. The `status` renderer calls this once per repo path
/// and shares the result with `run_with_theme_hex_prefetched`, so one
/// `git status` spawn serves the whole render instead of one per widget.
pub(crate) fn prefetch_porcelain(repo_path: &str) -> ParsedStatusV2 {
    get_porcelain_v2(repo_path)
}

/// Render the git status widget.
/// Shows sync icon + branch + changed/insertions/deletions/untracked/stash/conflicts/ahead/behind counts.
pub fn run(theme: Theme, repo_path: &str) -> String {
    let theme_hex = ThemeHex::from_theme(theme);
    let status = get_porcelain_v2(repo_path);
    run_with_theme_hex_prefetched(theme, &theme_hex, &status, repo_path)
}

/// Like `run_with_theme_hex`, but accepts a pre-parsed porcelain status so the
/// `status` renderer can share one `git status` spawn across widgets that run
/// in the same process for the same repo.
pub(crate) fn run_with_theme_hex_prefetched(
    theme: Theme,
    theme_hex: &ThemeHex,
    status: &ParsedStatusV2,
    repo_path: &str,
) -> String {
    let ctx = WidgetContext::from_theme_hex(theme, theme_hex);
    let theme = ctx.theme;

    let branch_raw = match status.branch.as_deref() {
        Some(b) if !b.is_empty() && b != "HEAD" => b,
        _ => return String::new(),
    };

    let display_branch = trim_branch_name(branch_raw);

    let untracked = status.untracked;
    let conflict_count = status.conflicts;
    let ahead = status.ahead;
    let behind = status.behind;
    let changed = status.changed;

    let sync_mode = if changed > 0 {
        SyncMode::Dirty
    } else if ahead > 0 {
        SyncMode::Ahead
    } else if behind > 0 {
        SyncMode::Behind
    } else {
        SyncMode::Clean
    };

    let (insertions, deletions) = if changed > 0 {
        get_diff_stats_cached(repo_path)
    } else {
        (0, 0)
    };
    let stash_count = status.stashes;

    let bg = theme_hex.color(theme.surface);
    let warning = theme_hex.color(theme.warning);
    let success = theme_hex.color(theme.success);
    let danger = theme_hex.color(theme.danger);
    let danger_bright = theme_hex.color(theme.danger_bright);
    let info_bright = theme_hex.color(theme.info_bright);
    let muted = theme_hex.color(theme.muted);

    let mut result = String::with_capacity(512);

    // Continuous surface block: one leading reset, then every sub-segment
    // (sync icon, branch, counts) paints on the same `surface` background.
    // No internal resets — those painted 1-cell `background` gaps between
    // sub-segments inside the block.
    result.push_str(ctx.reset.as_str());

    match sync_mode {
        SyncMode::Dirty => {
            let _ = write!(
                result,
                "#[bg={},fg={},bold]▒ 󱓎",
                bg, danger_bright
            );
        }
        SyncMode::Ahead => {
            let _ = write!(result, "#[bg={},fg={},bold]▒ 󰛃", bg, danger);
        }
        SyncMode::Behind => {
            let _ = write!(
                result,
                "#[bg={},fg={},bold]▒ 󰛀",
                bg, info_bright
            );
        }
        SyncMode::Clean => {
            let _ = write!(result, "#[bg={},fg={},bold]▒ ", bg, success);
        }
    }

    let _ = write!(
        result,
        " #[fg={},bg={},bold]{}",
        theme_hex.color(theme.foreground),
        bg,
        display_branch
    );

    if changed > 0 {
        let _ = write!(
            result,
            " #[fg={},bg={},bold] {}",
            warning, bg, changed
        );
    }

    if insertions > 0 {
        let _ = write!(
            result,
            " #[fg={},bg={},bold] {}",
            success, bg, insertions
        );
    }

    if deletions > 0 {
        let _ = write!(
            result,
            " #[fg={},bg={},bold] {}",
            danger, bg, deletions
        );
    }

    if untracked > 0 {
        let _ = write!(
            result,
            " #[fg={},bg={},bold] {}",
            muted, bg, untracked
        );
    }

    if stash_count > 0 {
        let _ = write!(
            result,
            " #[fg={},bg={},bold] {}",
            info_bright, bg, stash_count
        );
    }

    if conflict_count > 0 {
        let _ = write!(
            result,
            " #[fg={},bg={},bold]󰅘 {}",
            danger_bright, bg, conflict_count
        );
    }

    if ahead > 0 {
        let _ = write!(
            result,
            " #[fg={},bg={},bold]↑{}",
            info_bright, bg, ahead
        );
    }

    if behind > 0 {
        let _ = write!(
            result,
            " #[fg={},bg={},bold]↓{}",
            danger, bg, behind
        );
    }

    result.push(' ');
    result
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::themes;

    #[test]
    fn git_status_produces_valid_tmux_output_in_project_repo() {
        let output = run(themes::hard::THEME, ".");
        if !output.is_empty() {
            assert!(output.contains("#[fg=") || output.contains("#[bg="));
            // The '▒' separator glyph precedes the sync icon.
            assert!(output.contains("▒"));
            assert!(output.ends_with(' '));
        }
    }
}
